services/api/service.go

```diff
@@ -835,6 +924,71 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
 		if err != nil {
 			log.WithError(err).Error("failed to increment builder-stats after getPayload")
 		}
+
+		builderStatus, err := api.redis.GetBlockBuilderStatus(bidTrace.BuilderPubkey.String())
+		if err != nil {
+			log.WithError(err).Error("failed to get builder status")
+		}
+
+		// Check if the block was valid if the builder is OptimisticActive or OptimisticDemoted.
+		if builderStatus == common.OptimisticActive || builderStatus == common.OptimisticDemoted {
+			// Set to locked while we process the winning block.
+			api.setStatusByCollateralID(bidTrace.BuilderPubkey.String(), common.OptimisticLocked)
```

Comment:
(Self-TODO: think about what can go wrong with reorgs.)

Reply:
thinking about this a bit, i dont see any issue actually. once they call `getPayload` through us, that beacon block at slot `n` is published. even if they were to try and use a valid block for slot `n` to reorg the `n+1` slot block out, now they have signed two blocks at slot `n` which is a slashing condition. basically we know that once that block is out there, they can never get that reward back. also another nice aspect of handling the refunds manually is that we can sanity check that the slot we are refunding is empty. given it will take place O(hours) later, we can be confident that no reorg is going to change that.

Reply:
lmk if i am missing something here!